### ISSUE TITLE
Datatypes: neuro/parcellation/surface and neuro/surface

### ISSUE THREAD

**Issue by bacaron** [brainlife/brainlife#71]
Currently the neuro/parcellation/surface and the neuro/surface datatypes require a naming format of something like lh.annot.gii. However, this naming schema does not follow the naming schema provided by connectome workbench, and thus makes the files incompatible (information on appropriate naming schema: https://balsa.wustl.edu/about/fileTypes). For something like a surface annotation, the naming schema should be something like lh.annot.label.gii.

However, there's also the issue of incompatibility between connectome workbench naming and Freesurfer naming. For example, Freesurfer names that annotation file as lh.annot. It is quite simple to convert from Freesurfer to connectome workbench, so I believe it would be best to stick with the Freesurfer naming convention. Then any apps requiring connectome workbench can convert the appropriate files to the correct naming schema. Converting back to freesurfer naming schema for the outputs should also be quite simple.

TL;DR: the current naming schema does not provide compatibility and easy conversion to be used with connectome workbench.

**Comment by soichih** [comment]
Can't we solve the naming compatibility issue simply by each Apps renaming files (symlinking) to desired naming convention? Different data analysis toolset expects different naming, so I don't think we can come up with naming conventions that satisfy everything?

neuro/surface (*.gii) datatype is currently only used by fmriprep, and analyzePRF app, but neuro/parcellation/surface is used by a few other Apps.

If we are to replace these datatypes with something better, I am wondering if we could replace them with the following 2 new datatypes

> [func/surface-data] https://brainlife.io/datatype/5f78b377268f76598c29b27a
> [surface/vertices] https://brainlife.io/datatype/5f78b255268f764bdd29b254 

Below is a bit more information about these 2 new datatypes.

--------------------------------------

For the new surface/data DT (datatype), I thought we should register surface/vertices and func/surface-data DT separately.. The reason why I think we should separate surfaces and data is that there could be many surfaces(3d mesh) that could go along with a single data file. Actually, there could be many(data measures)-to-many(surface-vertices) relationships. I think creating 2 separate DT would allow us to organize them more easily.

To demonstrate this more clearly, I just registered 2 new DTs..

surface/vertices (3d-mesh) .surf.gii
> https://brainlife.io/datatype/5f78b255268f764bdd29b254
This stores various surface 3D meshes in a single directory. I think it makes sense to allow *many* surfaces stored inside a single DT as most surface generating Apps create more than just 1 surfaces. Left v.s. Right should be handled by having 2 separate objects of this DT with <left> or <right> datatype tag, so we can handle the case where the surface is not specific to either hemisphere.

func/surface-data (.data.gii)
https://brainlife.io/datatype/5f78b377268f76598c29b27a
This DL contains a single func.data.gii. If an App generates data for both lh and rh, then the App will need to output 2 separate objects. Each object can then mapped (via the provenance) to surface/vertices object to provide the 3D coordinates. Similar to surface/vertices, Left v.s. Right should be handled via datatype tag.

The file names used in these DTs can be symlinked or copied to more connectome_wb friendly names at run time - using datatype tags, or other metadata.

----------------------------------------

Not to muddy this conversation.. I recently learned that we could also store surface/vertices in cifti format. (.nii). Think we could have both surface/vertices-gii and surface/vertices-cifti so we can handle both format. verticies-cifti is normally used to store HCP dtseries (time series data) - which might be different from surface/verticies which only stores a single time frame, but I am not sure.

**Comment by bacaron** [comment]
Can't we solve the naming compatibility issue simply by each Apps renaming files (symlinking) to desired naming convention? Different data analysis toolset expects different naming, so I don't think we can come up with naming conventions that satisfy everything?

_This doesn't work. It's not just the names, but the assumptions of the structure that come with the name. A simple renaming of the file does not fix the issue and a more direct conversion of proper files is needed for it to work properly in connectome workbench._

neuro/surface (*.gii) datatype is currently only used by fmriprep, and analyzePRF app, but neuro/parcellation/surface is used by a few other Apps.

If we are to replace these datatypes with something better, I am wondering if we could replace them with the following 2 new datatypes

[func/surface-data] https://brainlife.io/datatype/5f78b377268f76598c29b27a
[surface/vertices] https://brainlife.io/datatype/5f78b255268f764bdd29b254

Below is a bit more information about these 2 new datatypes.

For the new surface/data DT (datatype), I thought we should register surface/vertices and func/surface-data DT separately.. The reason why I think we should separate surfaces and data is that there could be many surfaces(3d mesh) that could go along with a single data file. Actually, there could be many(data measures)-to-many(surface-vertices) relationships. I think creating 2 separate DT would allow us to organize them more easily.

__Overall i'm generally fine with storing the surface and metric files as separate datatypes. It will be sort of a pain to adjust the apps that currently output files like this (cortex mapping, prf fit, etc)._

To demonstrate this more clearly, I just registered 2 new DTs..

surface/vertices (3d-mesh) .surf.gii

https://brainlife.io/datatype/5f78b255268f764bdd29b254
This stores various surface 3D meshes in a single directory. I think it makes sense to allow many surfaces stored inside a single DT as most surface generating Apps create more than just 1 surfaces. Left v.s. Right should be handled by having 2 separate objects of this DT with or datatype tag, so we can handle the case where the surface is not specific to either hemisphere.

_I'm not a fan of having a set of files per hemisphere though. A few of my stats app work more efficiently with both hemispheres and adding redundant inputs with different tags just causes confusion imo. Happy to discuss more in person about this though._

func/surface-data (.data.gii)
https://brainlife.io/datatype/5f78b377268f76598c29b27a
This DL contains a single func.data.gii. If an App generates data for both lh and rh, then the App will need to output 2 separate objects. Each object can then mapped (via the provenance) to surface/vertices object to provide the 3D coordinates. Similar to surface/vertices, Left v.s. Right should be handled via datatype tag.

The file names used in these DTs can be symlinked or copied to more connectome_wb friendly names at run time - using datatype tags, or other metadata.

Not to muddy this conversation.. I recently learned that we could also store surface/vertices in cifti format. (.nii). Think we could have both surface/vertices-gii and surface/vertices-cifti so we can handle both format. verticies-cifti is normally used to store HCP dtseries (time series data) - which might be different from surface/verticies which only stores a single time frame, but I am not sure.

_Yeah we can definitely do dtseries or cifti, but those data formats are super obtuse and lots of people hate trying to work with cifti. It might be helpful to output both gifti and cifti/dtseries, like what I do in the HCP Cortex Mapping replication app (https://brainlife.io/app/5f6cc1836bbee392069a19a6)_

**Comment by soichih** [comment]
I've registered cifti datatype > https://brainlife.io/datatype/5fa9a60cafcdcc57e2c444ad The new version of fmriprep generates this for surface version @faskowit

**Comment by soichih** [comment]
> This doesn't work. It's not just the names, but the assumptions of the structure that come with the name. A simple renaming of the file does not fix the issue and a more direct conversion of proper files is needed for it to work properly in connectome workbench.

@bacaron Can you provide some concrete examples of this issue? Can you contrast between what we are currently writing out v.s. how those files needs to be mapped / reorged so that you can do processing via connectome_wb?

**Comment by soichih** [comment]
We've met during our datatype weekly meeting to further discuss surface (related) datatypes. We've agreed to deprecate the existing neuro/parcellation/surface datatype (https://brainlife.io/datatype/5dec20ff6c0bd9f84485779f) which contains various gii files (both data and surfaces) and replace it by the following new datatypes.

* neuro/surface/data (https://brainlife.io/datatype/5f78b377268f76598c29b27a) 
Surface data stored in gifti. This datatype only contains measures at each vertices. data is split into left.gii and right.gii and an optional label.json in case this contains data like parcellation (each voxel value means different parcellation)

* neuro/surface/verticies https://brainlife.io/datatype/5f78b255268f764bdd29b254
This datatype contains surface vertices coordinates for each surfaces. It can store any number of surfaces. 

Finally, we've also registered a new cifti datatype (https://brainlife.io/datatype/5fa9a60cafcdcc57e2c444ad) Unlike gifti, cifti can contain data for both surface and sub-cortical surfaces and the new version of fmriprep outputs surface data in this format. As cifti can store wide range of data, I have preregistered various datatype tags to ensure that Apps can pick the correct cifti input.

* dconn: a ‘dense’ by ‘dense’ matrix (typically a ‘dense connectome’ from resting-state fMRI)
* dscalar: one or more dense maps of scalar values (e.g., myelin maps, curvature maps)
* dtseries: one or more dense timeseries datasets (e.g., fMRI timeseries) or other data at equal intervals
* dlabel: one or more dense maps of integer values, plus a ‘label table’ that defines each integer (e.g., one or more parcellations of cerebral cortex, subcortical nuclei, or all grayordinates)
* dpconn: a parcel by dense matrix
* pconn: a parcel by parcel matrix (typically, a ‘parcellated connectome’)
* pdconn: a dense by parcel matrix
* pscalar: one or more parcellated scalar maps (e.g., a parcellated thickness map having uniform thickness within each parcel
* ptseries: parcellated timeseries
* plabel: one or more parcel maps, where each parcel is identified and colored using a label
* sdseries: contains rows of series data (frequency, time, etc.). This data type is different than most other CIFTI files, as it is not mapped to brainordinates (i.e., it has neither a dense nor a parcels dimension).  The data is viewed in its entirety as a matrix chart, or individual rows are viewed as a line chart.
* fiberTemp: contains orientations of dMRI fibers. This is actually the same format as a dscalar file (above), but the arrangement of maps in the file is fixed so that wb_view knows how to translate them into fiber orientations.
